proposals/lambdas-04-kinds.md

```diff
@@ -0,0 +1,203 @@
+  - can raise an error into the outer scope
+  - can return early, exiting the outer scope
+
+Ref cap restrictions:
+  - can capture non-sendable values from the outer scope
+  - can receive non-sendable yielded values
+  - can return a non-sendable yield result value
+
+Mobility:
+  - CANNOT cross an actor boundary
+  - CANNOT be objectified or stored away for later (yields must occur in the immediate scope of the called function)
+
+Theoretical "as if" modeling:
+  - It's as if the called function state were kept on the stack to pause and continue execution across multiple intermediate returns
+
+Internal Implementation / Optimization:
+  - the block has no reference capability to track, as it is not objectified
+  - zero or more calls with intermediate return values representing yields from the called function
+  - stack-allocated "continuation struct" persists stack state for called function across multiple calls
+  - for a self-recursive yielding function, the continuation struct must be heap-allocated instead of stack-allocated, but the rest of the semantics are not affected.
+
+## Kind B (single-use isolated lambda)
+
+Call count:
+  - can be called zero or one time
+  - CANNOT be called more than once
+
+Inter-scope interaction:
+  - can capture variables from the outer scope
+  - can rebind captured variables in the inner scope (but not the outer scope)
+  - can consume `iso` variables in the inner scope, provided that they aren't used any more in the outer scope (and thus the consume does not affect the outer scope)
+  - CANNOT affect variable bindings in the outer scope
+  - CANNOT raise an error into the outer scope
+  - CANNOT return early, exiting the outer scope
+
+Ref cap restrictions:
+  - CANNOT capture non-sendable values from the outer scope
+  - can receive non-sendable yielded values
+  - can return a non-sendable yield result value
+
+Mobility:
+  - can cross an actor boundary
+  - can be objectified and stored away for later, but only as an `iso` with `:fun iso` so as to be single-use
+
+Theoretical "as if" modeling:
+  - It's as if the block becomes an object (with ref cap: `iso`)
+  - the object has a single function corresponding to the block body
+  - the captured variables are arguments to a `:new iso` constructor
+  - within the block, the captured variables are fields, which may be rebound internally
+  - the block function is a `:fun iso` that immediately lowers itself to `ref`, so that the captured variable fields may be seen through the un-colored `ref` lens.
+  - the `ref` is not recovered to an `iso`, so the function is not callable again
+
+Internal Implementation / Optimization:
+  - when objectified, it is a heap-allocated object of an anonymous one-off type, behaving like the "as if" model, and using a virtual table for dispatch when called.
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
```

Comment:
Also I think a behavior call can never be inlined by LLVM, but it could definitely be specialized by my compiler ahead of time, so that's a common example of a need for me to take action on this, assuming that LLVM can only assist with eliminating virtual dispatch on inlined content (though again, I would be quite happy to hear about me being wrong about that)

Reply:
I'm going to guess that LLVM can prove that for more cases than what you've mentioned. Or to put another way, I imagine that getting any cases that LLVM doesn't cover is probably going to be more effort than is worth spending any time in the foreseeable future

Reply:
What specialization would occur that couldn't be performed at the LLVM level?

Reply:
I think the behavior call example I mentioned above is a common use case and reliable to reproduce that I don't think LLVM will be able to devirtualize.

If I have a behavior that accepts an abstract type as a parameter, and I pass it a concrete type as the corresponding argument, any calls made with the parameter as a receiver will be virtual calls and LLVM has no way to devirtualize them, because it isn't able to inline behavior dispatch into the behavior caller, and it has no other way to track the concrete descriptor from the behavior call site into the behavior body.

By contrast in the Savi compiler I think it would not be too tricky to generate a specialized version of the behavior that accepts the concrete type as a parameter and give it a totally distinct message id. LLVM doesn't know that it's okay to do this (make up new message ids and add corresponding new clauses to the dispatch function) so it won't, but I know the semantics of the Pony runtime and how I am using it, so I can encode the knowledge of this optimization into the Savi compiler.

There are perhaps other examples where pre-LLVM specialization can do a better job than LLVM can, but that's the clearest example I can come up with at the moment.